diff -Naur linux-4.20.6/fs/aufs/cpup.c linux-4.20.6.change/fs/aufs/cpup.c
--- linux-4.20.6/fs/aufs/cpup.c	2018-11-03 06:34:01.000000000 +0530
+++ linux-4.20.6.change/fs/aufs/cpup.c	2019-02-03 11:08:04.764371609 +0530
@@ -396,7 +396,7 @@
 	h_src_inode = file_inode(src);
 	h_src_sb = h_src_inode->i_sb;
 	if (h_src_sb != file_inode(dst)->i_sb
-	    || !dst->f_op->clone_file_range) {
+	    || !dst->f_op->copy_file_range) {
 		err = au_do_copy(dst, src, len);
 		goto out;
 	}
diff -Naur linux-4.20.6/fs/aufs/vfsub.h linux-4.20.6.change/fs/aufs/vfsub.h
--- linux-4.20.6/fs/aufs/vfsub.h	2018-11-03 06:34:12.000000000 +0530
+++ linux-4.20.6.change/fs/aufs/vfsub.h	2019-02-03 11:08:04.764371609 +0530
@@ -278,7 +278,7 @@
 	int err;
 
 	lockdep_off();
-	err = vfs_clone_file_range(src, 0, dst, 0, len);
+	err = vfs_clone_file_range(src, 0, dst, 0, len, 0);
 	lockdep_on();
 
 	return err;
