diff -Naur gimp-gtk3-port.a/libgimpcolor/gimpcolortransform.c gimp-gtk3-port.b/libgimpcolor/gimpcolortransform.c
--- gimp-gtk3-port.a/libgimpcolor/gimpcolortransform.c	2018-05-20 20:57:44.000000000 +0530
+++ gimp-gtk3-port.b/libgimpcolor/gimpcolortransform.c	2019-04-11 17:33:36.379285062 +0530
@@ -18,7 +18,7 @@
  *
  * You should have received a copy of the GNU Lesser General Public
  * License along with this library.  If not, see
- * <http://www.gnu.org/licenses/>.
+ * <https://www.gnu.org/licenses/>.
  */
 
 #include "config.h"
@@ -83,8 +83,8 @@
 static void   gimp_color_transform_finalize (GObject *object);
 
 
-G_DEFINE_TYPE (GimpColorTransform, gimp_color_transform,
-               G_TYPE_OBJECT);
+G_DEFINE_TYPE_WITH_PRIVATE (GimpColorTransform, gimp_color_transform,
+                            G_TYPE_OBJECT)
 
 #define parent_class gimp_color_transform_parent_class
 
@@ -131,17 +131,13 @@
                   G_TYPE_NONE, 1,
                   G_TYPE_DOUBLE);
 
-  g_type_class_add_private (klass, sizeof (GimpColorTransformPrivate));
-
   cmsSetLogErrorHandler (lcms_error_handler);
 }
 
 static void
 gimp_color_transform_init (GimpColorTransform *transform)
 {
-  transform->priv = G_TYPE_INSTANCE_GET_PRIVATE (transform,
-                                                 GIMP_TYPE_COLOR_TRANSFORM,
-                                                 GimpColorTransformPrivate);
+  transform->priv = gimp_color_transform_get_instance_private (transform);
 }
 
 static void
@@ -488,7 +484,7 @@
       iter = gegl_buffer_iterator_new (src_buffer, src_rect, 0,
                                        priv->src_format,
                                        GEGL_ACCESS_READ,
-                                       GEGL_ABYSS_NONE);
+                                       GEGL_ABYSS_NONE, 2);
 
       gegl_buffer_iterator_add (iter, dest_buffer, dest_rect, 0,
                                 priv->dest_format,
@@ -500,15 +496,15 @@
           if (priv->transform)
             {
               cmsDoTransform (priv->transform,
-                              iter->data[0], iter->data[1], iter->length);
+                              iter->items[0].data, iter->items[1].data, iter->length);
             }
           else
             {
               babl_process (priv->fish,
-                            iter->data[0], iter->data[1], iter->length);
+                            iter->items[0].data, iter->items[1].data, iter->length);
             }
 
-          done_pixels += iter->roi[0].width * iter->roi[0].height;
+          done_pixels += iter->items[0].roi.width * iter->items[0].roi.height;
 
           g_signal_emit (transform, gimp_color_transform_signals[PROGRESS], 0,
                          (gdouble) done_pixels /
@@ -520,22 +516,22 @@
       iter = gegl_buffer_iterator_new (src_buffer, src_rect, 0,
                                        priv->src_format,
                                        GEGL_ACCESS_READWRITE,
-                                       GEGL_ABYSS_NONE);
+                                       GEGL_ABYSS_NONE, 1);
 
       while (gegl_buffer_iterator_next (iter))
         {
           if (priv->transform)
             {
               cmsDoTransform (priv->transform,
-                              iter->data[0], iter->data[0], iter->length);
+                              iter->items[0].data, iter->items[0].data, iter->length);
             }
           else
             {
               babl_process (priv->fish,
-                            iter->data[0], iter->data[0], iter->length);
+                            iter->items[0].data, iter->items[0].data, iter->length);
             }
 
-          done_pixels += iter->roi[0].width * iter->roi[0].height;
+          done_pixels += iter->items[0].roi.width * iter->items[0].roi.height;
 
           g_signal_emit (transform, gimp_color_transform_signals[PROGRESS], 0,
                          (gdouble) done_pixels /
