diff -Naur a/util/grub-mkimagexx.c b/util/grub-mkimagexx.c
--- a/util/grub-mkimagexx.c	2018-10-21 13:48:05.660265234 +0530
+++ b/util/grub-mkimagexx.c	2018-10-21 14:04:32.007372338 +0530
@@ -832,6 +832,7 @@
 		  break;
 
 		case R_X86_64_PC32:
+		case R_X86_64_PLT32:
 		  {
 		    grub_uint32_t *t32 = (grub_uint32_t *) target;
 		    *t32 = grub_host_to_target64 (grub_target_to_host32 (*t32)
diff -Naur a/util/grub-module-verifier.c b/util/grub-module-verifier.c
--- a/util/grub-module-verifier.c	2018-10-21 13:48:05.659265235 +0530
+++ b/util/grub-module-verifier.c	2018-10-21 14:05:02.263386552 +0530
@@ -19,6 +19,7 @@
       -1
     }, (int[]){
       R_X86_64_PC32,
+      R_X86_64_PLT32,
       -1
     }
   },
