diff -Naur gst-plugins-bad-1.14.4.a/ext/fdkaac/gstfdkaacdec.c gst-plugins-bad-1.14.4.b/ext/fdkaac/gstfdkaacdec.c
--- gst-plugins-bad-1.14.4.a/ext/fdkaac/gstfdkaacdec.c	2019-01-15 08:02:18.282492992 +0530
+++ gst-plugins-bad-1.14.4.b/ext/fdkaac/gstfdkaacdec.c	2019-01-15 08:03:10.230068894 +0530
@@ -151,17 +151,18 @@
     gst_buffer_unref (codec_data);
   }
 
-  if ((err =
-          aacDecoder_SetParam (self->dec, AAC_PCM_OUTPUT_CHANNEL_MAPPING,
-              0)) != AAC_DEC_OK) {
-    GST_ERROR_OBJECT (self, "Failed to set output channel mapping: %d", err);
+  err = aacDecoder_SetParam (self->dec, AAC_PCM_MAX_OUTPUT_CHANNELS, 0);
+  if (err != AAC_DEC_OK) {
+    GST_ERROR_OBJECT (self, "Failed to disable downmixing: %d", err);
     return FALSE;
   }
 
-  if ((err =
-          aacDecoder_SetParam (self->dec, AAC_PCM_OUTPUT_INTERLEAVED,
-              1)) != AAC_DEC_OK) {
-    GST_ERROR_OBJECT (self, "Failed to set interleaved output: %d", err);
+  /* Choose WAV channel mapping to get interleaving even with libfdk-aac 2.0.0
+   * The pChannelIndices retain the indices from the standard MPEG mapping so
+   * we're agnostic to the actual order. */
+  err = aacDecoder_SetParam (self->dec, AAC_PCM_OUTPUT_CHANNEL_MAPPING, 1);
+  if (err != AAC_DEC_OK) {
+    GST_ERROR_OBJECT (self, "Failed to set output channel mapping: %d", err);
     return FALSE;
   }
 
@@ -194,28 +195,27 @@
     gst_buffer_map (inbuf, &imap, GST_MAP_READ);
     valid = size = imap.size;
 
-    if ((err =
-            aacDecoder_Fill (self->dec, (guint8 **) & imap.data, &size,
-                &valid)) != AAC_DEC_OK) {
+    err = aacDecoder_Fill (self->dec, (guint8 **) & imap.data, &size, &valid);
+    if (err != AAC_DEC_OK) {
       GST_AUDIO_DECODER_ERROR (self, 1, STREAM, DECODE, (NULL),
           ("filling error: %d", err), ret);
       goto out;
     }
 
-    if (GST_BUFFER_IS_DISCONT (inbuf))
+    if (GST_BUFFER_IS_DISCONT (inbuf)) {
       flags |= AACDEC_INTR;
+    }
   } else {
     flags |= AACDEC_FLUSH;
   }
 
-  if ((err =
-          aacDecoder_DecodeFrame (self->dec, self->decode_buffer,
-              self->decode_buffer_size, flags)) != AAC_DEC_OK) {
-    if (err == AAC_DEC_TRANSPORT_SYNC_ERROR) {
-      ret = GST_FLOW_OK;
-      outbuf = NULL;
-      goto finish;
-    }
+  err = aacDecoder_DecodeFrame (self->dec, self->decode_buffer,
+      self->decode_buffer_size, flags);
+  if (err == AAC_DEC_TRANSPORT_SYNC_ERROR) {
+    ret = GST_FLOW_OK;
+    outbuf = NULL;
+    goto finish;
+  } else if (err != AAC_DEC_OK) {
     GST_AUDIO_DECODER_ERROR (self, 1, STREAM, DECODE, (NULL),
         ("decoding error: %d", err), ret);
     goto out;
@@ -232,14 +232,12 @@
   if (stream_info->numChannels == 1) {
     pos[0] = GST_AUDIO_CHANNEL_POSITION_MONO;
   } else {
-    gint n_front = 0, n_side = 0, n_back = 0, n_lfe = 0;
+    gint n_front = 0, n_back = 0, n_lfe = 0;
 
     /* FIXME: Can this be simplified somehow? */
     for (i = 0; i < stream_info->numChannels; i++) {
       if (stream_info->pChannelType[i] == ACT_FRONT) {
         n_front++;
-      } else if (stream_info->pChannelType[i] == ACT_SIDE) {
-        n_side++;
       } else if (stream_info->pChannelType[i] == ACT_BACK) {
         n_back++;
       } else if (stream_info->pChannelType[i] == ACT_LFE) {
@@ -303,37 +301,50 @@
           ret = GST_FLOW_NOT_NEGOTIATED;
           goto out;
         }
-      } else if (stream_info->pChannelType[i] == ACT_SIDE) {
-        if (n_side & 1) {
-          GST_ERROR_OBJECT (self, "Odd number of side channels not supported");
-          ret = GST_FLOW_NOT_NEGOTIATED;
-          goto out;
-        } else if (stream_info->pChannelIndices[i] == 0) {
-          pos[i] = GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT;
-        } else if (stream_info->pChannelIndices[i] == 1) {
-          pos[i] = GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT;
-        } else {
-          GST_ERROR_OBJECT (self, "Side channel index %d not supported",
-              stream_info->pChannelIndices[i]);
-          ret = GST_FLOW_NOT_NEGOTIATED;
-          goto out;
-        }
       } else if (stream_info->pChannelType[i] == ACT_BACK) {
         if (stream_info->pChannelIndices[i] == 0) {
           if (n_back & 1)
             pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_CENTER;
+          else if (n_back > 2)
+            pos[i] = GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT;
           else
             pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_LEFT;
         } else if (stream_info->pChannelIndices[i] == 1) {
-          if (n_back & 1)
+          if ((n_back & 1) && n_back > 3)
+            pos[i] = GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT;
+          else if (n_back & 1)
             pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_LEFT;
+          else if (n_back > 2)
+            pos[i] = GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT;
           else
             pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT;
         } else if (stream_info->pChannelIndices[i] == 2) {
-          if (n_back & 1)
+          if ((n_back & 1) && n_back > 3)
+            pos[i] = GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT;
+          else if (n_back & 1)
+            pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT;
+          else if (n_back > 2)
+            pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_LEFT;
+          else
+            g_assert_not_reached ();
+        } else if (stream_info->pChannelIndices[i] == 3) {
+          if ((n_back & 1) && n_back > 3)
+            pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_LEFT;
+          else if (n_back & 1)
+            g_assert_not_reached ();
+          else if (n_back > 2)
             pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT;
           else
             g_assert_not_reached ();
+        } else if (stream_info->pChannelIndices[i] == 4) {
+          if ((n_back & 1) && n_back > 3)
+            pos[i] = GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT;
+          else if (n_back & 1)
+            g_assert_not_reached ();
+          else if (n_back > 2)
+            g_assert_not_reached ();
+          else
+            g_assert_not_reached ();
         } else {
           GST_ERROR_OBJECT (self, "Side channel index %d not supported",
               stream_info->pChannelIndices[i]);
@@ -409,9 +420,9 @@
 
   if (self->dec) {
     AAC_DECODER_ERROR err;
-    if ((err =
-            aacDecoder_DecodeFrame (self->dec, self->decode_buffer,
-                self->decode_buffer_size, AACDEC_FLUSH)) != AAC_DEC_OK) {
+    err = aacDecoder_DecodeFrame (self->dec, self->decode_buffer,
+        self->decode_buffer_size, AACDEC_FLUSH);
+    if (err != AAC_DEC_OK) {
       GST_ERROR_OBJECT (self, "flushing error: %d", err);
     }
   }
@@ -448,3 +459,4 @@
   GST_DEBUG_CATEGORY_INIT (gst_fdkaacdec_debug, "fdkaacdec", 0,
       "fdkaac decoder");
 }
+
diff -Naur gst-plugins-bad-1.14.4.a/ext/fdkaac/gstfdkaacenc.c gst-plugins-bad-1.14.4.b/ext/fdkaac/gstfdkaacenc.c
--- gst-plugins-bad-1.14.4.a/ext/fdkaac/gstfdkaacenc.c	2019-01-15 08:02:18.282492992 +0530
+++ gst-plugins-bad-1.14.4.b/ext/fdkaac/gstfdkaacenc.c	2019-01-15 08:03:53.060744646 +0530
@@ -62,56 +62,110 @@
   GstAudioChannelPosition positions[8];
 } channel_layouts[] = {
   {
+    /* MPEG 1: Mono */
     1, MODE_1, {
   GST_AUDIO_CHANNEL_POSITION_MONO}}, {
+    /* MPEG 2: Stereo */
     2, MODE_2, {
   GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
           GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT}}, {
+    /* MPEG 3: Stereo + Center */
     3, MODE_1_2, {
   GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
           GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
           GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT}}, {
-    3, MODE_2_1, {
-  GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
-          GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
-          GST_AUDIO_CHANNEL_POSITION_LFE1}}, {
+    /* MPEG 4: Stereo + Center + Rear center */
     4, MODE_1_2_1, {
   GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
           GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
           GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
           GST_AUDIO_CHANNEL_POSITION_REAR_CENTER}}, {
+    /* MPEG 5: 5.0 Surround */
     5, MODE_1_2_2, {
-  GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
+      GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
           GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
           GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
-          GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT,
-          GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT}}, {
+          /* FIXME: SURROUND instead of REAR is more to spec */
+  GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT}}, {
+    /* MPEG 6: 5.1 Surround */
     6, MODE_1_2_2_1, {
-  GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
+      GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
           GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
           GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
-          GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT,
-          GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT,
+          /* FIXME: SURROUND instead of REAR is more to spec */
+  GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT,
+          GST_AUDIO_CHANNEL_POSITION_LFE1}}, {
+    /* MPEG 7: SDDS for cinema */
+    8, MODE_1_2_2_2_1, {
+      GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT_OF_CENTER,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT_OF_CENTER,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
+          /* FIXME: SURROUND instead of REAR is more to spec */
+  GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT,
           GST_AUDIO_CHANNEL_POSITION_LFE1}}
+#ifdef HAVE_FDK_AAC_2_0_0
+  , {
+    /* MPEG 11: 6.1 Surround */
+    7, MODE_6_1, {
+      GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
+          /* FIXME: SURROUND instead of REAR is more to spec */
+  GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_CENTER,
+          GST_AUDIO_CHANNEL_POSITION_LFE1}}, {
+    /* MPEG 12: 7.1 Surround */
+    8, MODE_7_1_BACK, {
+      GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
+          /* FIXME: SURROUND instead of SIDE is more to spec */
+  GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT,
+          GST_AUDIO_CHANNEL_POSITION_LFE1}}, {
+    /* MPEG 14: 5.1.2 Surround */
+    8, MODE_7_1_TOP_FRONT, {
+      GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
+          /* FIXME: SURROUND instead of REAR is more to spec */
+  GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT,
+          GST_AUDIO_CHANNEL_POSITION_LFE1,
+          GST_AUDIO_CHANNEL_POSITION_TOP_FRONT_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_TOP_FRONT_RIGHT}}
+#endif
 #ifdef HAVE_FDK_AAC_0_1_4
   , {
+    /* Non-standard PCE clone of mode 12 */
     8, MODE_7_1_REAR_SURROUND, {
-  GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
+      GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
           GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
           GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
-          GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT,
+          /* FIXME: SURROUND instead of SIDE is more to spec */
+  GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT,
           GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT,
           GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
           GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT,
           GST_AUDIO_CHANNEL_POSITION_LFE1}}, {
+    /* Non-standard PCE clone of mode 7 */
     8, MODE_7_1_FRONT_CENTER, {
-  GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
+      GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
           GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT_OF_CENTER,
           GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT_OF_CENTER,
           GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
           GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
-          GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT,
-          GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT,
+          /* FIXME: SURROUND instead of REAR is more to spec */
+  GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
+          GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT,
           GST_AUDIO_CHANNEL_POSITION_LFE1}}
 #endif
 };
@@ -292,9 +346,8 @@
   if (allowed_caps)
     gst_caps_unref (allowed_caps);
 
-  if ((err =
-          aacEncOpen (&self->enc, 0,
-              GST_AUDIO_INFO_CHANNELS (info))) != AACENC_OK) {
+  err = aacEncOpen (&self->enc, 0, GST_AUDIO_INFO_CHANNELS (info));
+  if (err != AACENC_OK) {
     GST_ERROR_OBJECT (self, "Unable to open encoder: %d\n", err);
     return FALSE;
   }
@@ -529,11 +582,10 @@
   out_desc.bufSizes = &out_sizes;
   out_desc.bufElSizes = &out_el_sizes;
 
-  if ((err = aacEncEncode (self->enc, &in_desc, &out_desc, &in_args,
-              &out_args)) != AACENC_OK) {
-    if (!inbuf && err == AACENC_ENCODE_EOF)
-      goto out;
-
+  err = aacEncEncode (self->enc, &in_desc, &out_desc, &in_args, &out_args);
+  if (err == AACENC_ENCODE_EOF && !inbuf)
+    goto out;
+  else if (err != AACENC_OK) {
     GST_ERROR_OBJECT (self, "Failed to encode data: %d", err);
     ret = GST_FLOW_ERROR;
     goto out;
@@ -612,3 +664,4 @@
   GST_DEBUG_CATEGORY_INIT (gst_fdkaacenc_debug, "fdkaacenc", 0,
       "fdkaac encoder");
 }
+
