diff -Naur a/src/idevice.c b/src/idevice.c
--- a/src/idevice.c	2015-01-28 06:40:32.000000000 +0530
+++ b/src/idevice.c	2018-05-22 03:04:31.758732015 +0530
@@ -678,7 +678,7 @@
 	}
 	BIO_set_fd(ssl_bio, (int)(long)connection->data, BIO_NOCLOSE);
 
-	SSL_CTX *ssl_ctx = SSL_CTX_new(SSLv3_method());
+	SSL_CTX *ssl_ctx = SSL_CTX_new(SSLv23_method());
 	if (ssl_ctx == NULL) {
 		debug_info("ERROR: Could not create SSL context.");
 		BIO_free(ssl_bio);
